Documentation for incrementally consuming partitions of a partitioned file set.

anwar6953 opened the pull request:
Documentation for incrementally consuming partitions of a partitioned file set.
Additionally, minor updates to existing PartitionedFileSet documentation.

http://builds.cask.co/browse/CDAP-RBT341-1
https://issues.cask.co/browse/CDAP-2993

Two comments from previous PR (accidentally opened against develop branch):
https://github.com/caskdata/cdap/pull/3212#discussion_r34857840
https://github.com/caskdata/cdap/pull/3212#discussion_r34857891

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -162,12 +162,12 @@ Yet that can become tedious to manage, especially if the naming convention shoul
+file(s) at that path uniquely through its partition keys, or more broadly through conditions over the
+partition keys. For example, the months of February through June of a particular year, or the month of
+November in any year. By inheriting the attributes |---| such as format and schema |---| of FileSets,
+PartitionedFileSets are a powerful abstraction over data that is organized into files.
```
I think this paragraph is running on a bit too much, and the use of "for example" twice confusing. Perhaps simplify the first sentence by removing that part:
```
it associates a partition key with a path.
```

and

"file(s) at that path uniquely through its partition keys," -> "file(s) at that path uniquely through their partition keys,"

anwar6953 replied:
I changed 'its' to 'their' to address the second part of this comment, but in regards to the first part of this comment:

It doesn't seem to make sense to remove 'it associates a partition key with a path.'. Can you clarify?

awholegunch replied:
Sorry! I meant "remove the 'for example' part", leaving:

```
it associates a partition key with a path.
```

anwar6953 replied:
Ok, I removed: "; for example the year and month associated with that path"

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+    super.onFinish(succeeded, context);
+  }
+
+A limitation of the BatchPartitionConsumer is that there can not be concurrent runs of the MapReduce job.
```
```
A limitation of the ``BatchPartitionConsumer`` is that there can't be concurrent runs of the MapReduce job.
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
     "dataset.totals.output.partition.key.league" : "nfl"
   }
 
+Incrementally Processing Partitions of a PartitionedFileSets in MapReduce
```
Incrementally Processing Partitions of a PartitionedFileSets

The title is otherwise too long, and as the lead says, this is just one way to process a fileset...

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
   }
 
+Incrementally Processing Partitions of a PartitionedFileSets in MapReduce
+=========================================================================
```
Shorten to match title

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+Incrementally Processing Partitions of a PartitionedFileSets in MapReduce
+=========================================================================
+
+One way to process a partitioned file set is with a MapReduce program that runs repeatedly and,
```
One way to process a partitioned fileset is with a repeatedly-running MapReduce program that,

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+=========================================================================
+
+One way to process a partitioned file set is with a MapReduce program that runs repeatedly and,
+in every run, reads the partitions that have been added since its previous run. This requires
```
in each run, reads all partitions that have been added since its previous run.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+
+One way to process a partitioned file set is with a MapReduce program that runs repeatedly and,
+in every run, reads the partitions that have been added since its previous run. This requires
+that the MapReduce program persists between runs what partitions have already been consumed.
```
which partitions have already been consumed.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+One way to process a partitioned file set is with a MapReduce program that runs repeatedly and,
+in every run, reads the partitions that have been added since its previous run. This requires
+that the MapReduce program persists between runs what partitions have already been consumed.
+An easy way to do that is the BatchPartitionConsumer, an experimental feature introduced in CDAP 3.1.0.
```
```
An easy way is to use the ``BatchPartitionConsumer``, an experimental feature introduced in CDAP 3.1.0.
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+that the MapReduce program persists between runs what partitions have already been consumed.
+An easy way to do that is the BatchPartitionConsumer, an experimental feature introduced in CDAP 3.1.0.
+The MapReduce program is responsible for extending this abstract consumer class with methods to
+persist and the read back its state. In the following example, the state is persisted to a row in a
```
persist and then read back its state. In this example, the state is persisted to a row in a KeyValue Table; however, other types of Datasets can be used::

anwar6953 replied:
I took your suggestion here, except I left the 'also' after 'Datasets'. It seems like that emphasizes the fact that other datasets can be used in addition to KV Table.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+in every run, reads the partitions that have been added since its previous run. This requires
+that the MapReduce program persists between runs what partitions have already been consumed.
+An easy way to do that is the BatchPartitionConsumer, an experimental feature introduced in CDAP 3.1.0.
+The MapReduce program is responsible for extending this abstract consumer class with methods to
```
Your MapReduce program is responsible for extending this abstract consumer class with methods to

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+
+Then, in the ``beforeSubmit()`` method of the MapReduce, specify the partitioned file set to be used as input::
+
+  @Override
```
indent two spaces

anwar6953 replied:
indented.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+      PartitionedFileSet inputRecords =
+        batchPartitionConsumer.getPartitionedFileSetForConsuming(context, "inputRecords");
+      context.setInput("inputRecords", inputRecords);
+
```
as this is not the end of the method, add:
```
   ...
}
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,58 @@ For example, give these arguments when starting the MapReduce through a RESTful
+This will read back the previously persisted state, determine the new partitions to read based upon this
+state, and compute a new state to store in memory until a call to ``persist()``. The dataset it returns
+is instantiated with the set of new partitions to read as input.
+To save the state of partition processing, call the consumer's ``persist()`` method. This ensures that the
```
Insert a blank line before this line to create a new paragraph.

anwar6953 commented:
@awholegunch Addressed comments, with one question: https://github.com/caskdata/cdap/pull/3228#discussion_r35146334

@anew Please take a look at the method renamings:
https://github.com/caskdata/cdap/commit/1449d301288369eb7819d8a721bf13005959aac9

awholegunch commented:
Added a comment.

awholegunch commented:
One minor nit, otherwise LGTM.

anew commented on cdap-docs/developers-manual/source/building-blocks/datasets/fileset.rst:
```diff
@@ -317,6 +317,61 @@ For example, give these arguments when starting the MapReduce through a RESTful
+    super.onFinish(succeeded, context);
+  }
+
+A limitation of the ``BatchPartitionConsumer`` is that there can't be concurrent runs of the MapReduce job.
```
If you persist in beforeSubmit(), then you can have concurrent runs, right?

anwar6953 replied:
This doesn't allow for rollbacks upon failure: (assuming a mapreduce run kicks off after the current one's beforeSubmit, but before it fails).

anew replied:
yes, true, then you would have to retry after failure. OK. for now, let's leave it at this.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/BatchPartitionConsumer.java:
```diff
@@ -55,7 +58,8 @@
    * the implementing readBytes(DatasetContext) method.
    */
-  public PartitionedFileSet getPartitionedFileSet(DatasetContext datasetContext, String partitionedFileSetName) {
+  public PartitionedFileSet getPartitionedFileSetForConsuming(DatasetContext datasetContext,
```
I am still not a fan of this method name... let's brainstorm....
perhaps getConfiguredDataset()? Or configureAndGetDataset()? or just configureDataset()?

anwar6953 replied:
getConfiguredDataset() sounds good to me.

anew replied:
ok, let's do that

anwar6953 commented:
updated.

anew commented:
LGTM
